### DISCUSSION TITLE
Support OAuth2 JWT tokens for calls to subscribers

### DISCUSSION THREAD

**Discussion by VladimirIordanov** [svix/svix-webhooks#644] [Ideas]
For security reason we require a valid JWT token for all incoming requests to our microservices. According to OAuth2 flow all tokens have the expiration timestamp. So we need to generate the new token if the current is expired. For the moment Svix doesn't support a functionality for generating a token by provided secrets. It would be nice to have such option in Svix ["Create Endpoint" API](https://api.svix.com/docs#tag/Endpoint/operation/create_endpoint_api_v1_app__app_id__endpoint__post)

**Comment by tasn** [comment]
Thanks @VladimirIordanov for suggesting this!

I'd love to hear more about how you envision this to work exactly! Is it something like:
1. Set the OAuth2 secret and authentication URL for an endpoint.
2. Svix will check the expiration of the OAuth2 token and if expires get a new one using the secret.
3. Svix uses the newly generated token to make API calls?

**Reply by VladimirIordanov** [reply]
You got everything right. In my understanding for this functionality the payload of "Create endpoint" API can be extended by these parameters:
```
{
  authorization: {
     type: 'None' | 'OAuth2',
     tokenEndpoint: {
         grantType: 'grantType',
         url: 'https://my_token/enpoint?parameter1=value1&parameter2=value2',
         headers: {
           Content-Type: application/x-www-form-urlencoded
         },
         method: POST,
         body: 'grant_type=client_credentials&scope=_supported_scope_',
         resultTokenFieldName: 'access_token'
     }   
  }
}
```
Grant type is one [of these values](https://oauth.net/2/grant-types/). Content-type header is usually 'application/x-www-form-urlencoded' for the most systems. HTTP method is usually POST as well.

What do you think? Will it work? Or I missed something?

**Reply by tasn** [reply]
I think it all sounds reasonable. I just don't remember where the secret is stored. Is it in the URL? (so the URL is secret)
I really need to brush up my memory of OAuth2.

**Reply by VladimirIordanov** [reply]
I guess it depends on implementation of token endpoint. 
For AWS and Azure it is passed in body:
`client_id=_client_id_&scope=_scope_&client_secret=_secret_&grant_type=client_credentials`

For Auth0 it is passed in such format:
`audience=API_IDENTIFIER&grant_type=client_credentials&client_id=YOUR_CLIENT_ID&client_secret=YOUR_CLIENT_SECRET`

**Reply by tasn** [reply]
Ahhh, sorry, I missed the body one.

OK, so `body` and `url` are both secretish. I guess the whole `tokenEndpoint` can be secret.

Could you please open a ticket and refer to this discussion? We would love to add it to the roadmap!

**Reply by VladimirIordanov** [reply]
Great! Created an issue #645.

**Reply by tasn** [reply]
Thanks a lot!
